src/Conversion/ONNXToKrnl/Tensor/Shape.cpp

```diff
@@ -0,0 +1,67 @@
+//===----------------Shape.cpp - Lowering Shape Op----------------------=== //
+//
+// Copyright 2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Shape Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+
+using namespace mlir;
+
+struct ONNXShapeOpLowering : public ConversionPattern {
+  ONNXShapeOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXShapeOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXShapeOpAdaptor operandAdaptor(operands);
+    ONNXShapeOp shapeOp = llvm::cast<ONNXShapeOp>(op);
+    bool insertDealloc = checkInsertDealloc(op);
+    Location loc = op->getLoc();
+    // Get input data.
+    Value data = operandAdaptor.data();
+    ArrayRef<int64_t> dataShape = data.getType().cast<MemRefType>().getShape();
+    unsigned dataRank = dataShape.size();
+
+    Value resultOperand = shapeOp.shape();
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+
+    Value alloc;
+    if (hasAllConstantDimensions(outputMemRefType))
+      alloc =
+          insertAllocAndDealloc(outputMemRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDealloc(
+          outputMemRefType, loc, rewriter, insertDealloc, {resultOperand});
+
+    // Iterate along the data shape storing dim value to result.
+    Value storeVal;
+    for (int i = 0; i < dataRank; i++) {
+      // Create store index value.
+      Value storeIndex =
+          emitConstantOp(rewriter, loc, rewriter.getIndexType(), i);
+
+      // Checking for dynamic dimensions.
+      if (dataShape[i] == -1) {
+        Value shapeVal = rewriter.create<DimOp>(loc, data, i);
```

Comment:
Maybe if you change `i` to `storeIndex` the extra `DIMIND` will go away? If this is the case, then we might want to rename `storeIndex` to something like `dimIndex`.